File: streamlit-app/backend/resume_parser.py
"""
Resume file text extraction for the AI Resume Screener.

Converts an uploaded resume file (.txt, .pdf, or .docx) into plain text
that the analyzer can consume. Keeping this out of the Streamlit layer
preserves the UI/domain separation described in `app.py`.
"""

import io

from docx import Document
from pypdf import PdfReader

SUPPORTED_EXTENSIONS = ["txt", "pdf", "docx"]


class ResumeParseError(Exception):
    """Raised when an uploaded resume file cannot be converted to text."""


def _extract_pdf(data: bytes) -> str:
    """Extract text from every page of a PDF."""
    try:
        reader = PdfReader(io.BytesIO(data))
        pages = [page.extract_text() or "" for page in reader.pages]
    except Exception as exc:  # noqa: BLE001 - pypdf raises varied types
        raise ResumeParseError(f"Could not read PDF: {exc}") from exc
    return "\n\n".join(pages)


def _extract_docx(data: bytes) -> str:
    """Extract text from a .docx file's paragraphs and tables."""
    try:
        document = Document(io.BytesIO(data))
    except Exception as exc:  # noqa: BLE001 - python-docx raises varied types
        raise ResumeParseError(f"Could not read Word document: {exc}") from exc

    lines = [paragraph.text for paragraph in document.paragraphs]
    # Many resume templates lay out sections in tables.
    for table in document.tables:
        for row in table.rows:
            lines.append(" | ".join(cell.text.strip() for cell in row.cells))
    return "\n".join(lines)


def extract_resume_text(filename: str, data: bytes) -> str:
    """Convert an uploaded resume file into plain text.

    Args:
        filename: The uploaded file's name, used to detect its type.
        data: The raw file bytes.

    Returns:
        The extracted text, stripped of surrounding whitespace.

    Raises:
        ResumeParseError: If the file type is unsupported, the file is
            unreadable, or it contains no extractable text (e.g. a
            scanned image-only PDF).
    """
    extension = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""

    if extension == "txt":
        text = data.decode("utf-8", errors="ignore")
    elif extension == "pdf":
        text = _extract_pdf(data)
    elif extension == "docx":
        text = _extract_docx(data)
    else:
        raise ResumeParseError(
            f"Unsupported file type '{filename}'. "
            f"Upload one of: {', '.join(SUPPORTED_EXTENSIONS)}."
        )

    text = text.strip()
    if not text:
        raise ResumeParseError(
            "No text could be extracted from this file. If it's a scanned "
            "PDF, paste the resume text instead."
        )
    return text

File: lib/resumeParser.ts
import mammoth from "mammoth";
import { extractText, getDocumentProxy } from "unpdf";

import { clampToInputLimit } from "@/lib/limits";

/**
 * Converts an uploaded resume file (.txt, .pdf, .docx) into plain text.
 * Runs server-side only: both parsers are Node libraries.
 */

export const SUPPORTED_EXTENSIONS = ["txt", "pdf", "docx"] as const;
export const MAX_FILE_BYTES = 5 * 1024 * 1024;

export class ResumeParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ResumeParseError";
  }
}

async function extractPdf(data: Uint8Array): Promise<string> {
  try {
    const pdf = await getDocumentProxy(data);
    const { text } = await extractText(pdf, { mergePages: true });
    return text;
  } catch (error) {
    console.error("PDF parse failed:", error);
    throw new ResumeParseError(
      "That PDF couldn't be read. It may be corrupt or password-protected — try another file, or paste the text.",
    );
  }
}

async function extractDocx(data: Uint8Array): Promise<string> {
  try {
    const { value } = await mammoth.extractRawText({ buffer: Buffer.from(data) });
    return value;
  } catch (error) {
    console.error("Word parse failed:", error);
    throw new ResumeParseError(
      "That Word document couldn't be read. Try re-saving it as .docx, or paste the text.",
    );
  }
}

export async function extractResumeText(
  filename: string,
  data: Uint8Array,
): Promise<string> {
  if (data.byteLength > MAX_FILE_BYTES) {
    throw new ResumeParseError("That file is larger than 5 MB. Upload a smaller file.");
  }

  const extension = filename.includes(".")
    ? filename.split(".").pop()!.toLowerCase()
    : "";

  let text: string;
  if (extension === "txt") {
    text = new TextDecoder().decode(data);
  } else if (extension === "pdf") {
    text = await extractPdf(data);
  } else if (extension === "docx") {
    text = await extractDocx(data);
  } else {
    throw new ResumeParseError(
      `Unsupported file type '${filename}'. Upload one of: ${SUPPORTED_EXTENSIONS.join(", ")}.`,
    );
  }

  const trimmed = text.trim();
  if (!trimmed) {
    throw new ResumeParseError(
      "No text could be extracted from this file. If it's a scanned PDF, paste the resume text instead.",
    );
  }
  // A 30-page PDF can exceed what /api/analyze accepts; clamp here rather
  // than let the user hit a length error on text this app extracted.
  return clampToInputLimit(trimmed);
}
